tools/flowc/backends/nim/runtime.nim

```diff
@@ -8,68 +8,341 @@ import hashes
 import asyncdispatch
 import osproc
 import macros
+import system/widestrs
 import "flow_lib/httpServer_type"
 
 # Runtime for NIM backend
 
 {.experimental: "overloadableEnums".}
 
-proc rt_runtime_error*(what: string): void =
-  assert(false, "runtime error: " & what)
-
-proc rt_escape(s: string): string = 
-  var r: string = ""
-  for ch in s:
-    case ch:
-    of '\n': r.add("\\n")
-    of '\t': r.add("\\t")
-    of '\r': r.add("\\u000d")
-    of '\\': r.add("\\\\")
-    of '"':  r.add("\\\"")
-    else: r.add(ch)
-  return r
+#[ Flow string: either a UTF-16 string of UTF-8 ]#
+
+const use16BitString = when defined UTF8: false else: true
+
+when use16BitString:
+  type String = seq[Utf16Char]
+else:
+  type String = string
+
+proc rt_utf8_to_string*(s: string): String =
+  when use16BitString:
+    # Convert utf8 to utf16
+    var wcs = newWideCString(s)
+    result = newSeq[Utf16Char](wcs.len)
+    for i in 0 ..< wcs.len:
+      result[i] = wcs[i]
+  else: return s
+
+template `==`*(a: Utf16Char, b: Utf16Char): bool = cast[uint16](a) == cast[uint16](b)
+template `!=`*(a: Utf16Char, b: Utf16Char): bool = cast[uint16](a) != cast[uint16](b)
+template `<`*(a: Utf16Char, b: Utf16Char): bool = cast[uint16](a) < cast[uint16](b)
+template `<=`*(a: Utf16Char, b: Utf16Char): bool = cast[uint16](a) <= cast[uint16](b)
+template `>`*(a: Utf16Char, b: Utf16Char): bool = cast[uint16](a) > cast[uint16](b)
+template `>=`*(a: Utf16Char, b: Utf16Char): bool = cast[uint16](a) >= cast[uint16](b)
+
+proc rt_empty_string*(): String = return when use16BitString: @[] else: ""
+
+template rt_concat_strings*(s1: String, s2: String): String =
+  when use16BitString:
+    sequtils.concat(s1, s2)
+  else:
+    s1 & s2
+
+template rt_string_len*(s: String): int32 =
+  when use16BitString: int32(s.len) else: int32(s.runeLen)
+
+template rt_string_char_code*(s: String, i: int32): int32 =
+  when use16BitString: int32(cast[uint16](s[i])) else: int32(s.runeAtPos(i))
```

Comment:
I think for utf8 strings the flow getCharCodeAt function should just return the 8bit value at the index (including surrogates). 

This way it is more like utf16 where it also return surrogates. I suspect this will make more flow code work.